### PR TITLE
Change diffusion dislocation to log iteration

### PR DESCRIPTION
**Pull request by lhy11009**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

I changed the iteration scheme in the diffusion dislocation material model to use the log value of stress and strain. This is part of the effort to change the iteration scheme of the non-linear rheology in material models.
FYI @bobmyhill @naliboff.
In this PR, I added new functions named "compute_log_strain_rate_and_derivative" to both the diffusion and dislocation modules, while the old function of "compute_strain_rate_and_derivative is still kept.

I will need some help with one test zero_matrix, which is likely to fail (while I run in the standard docker image).  That one doesn't use the diffusion dislocation material model.  I'd like to first discuss this with Reviewers before merging this PR. But other than that, if this seems fine, we can first start the tester and debug this one failing test from there.

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

**Comment by bobmyhill** [comment]
Hi @lhy11009, thanks for trying this! 

Do you see any improvement in convergence behaviour in any tests\*? We should avoid changing functionality unless it provides a new feature / fixes a bug / produces a demonstrable improvement in performance / cleans up code.

This PR currently also changes user-facing functionality (via the change in the definition of the residual variable). The Peierls creep PR did this too, but I'm more concerned about the diffusion-dislocation model because it is used in many places in the code. Do we need to change this? If so, you should document it in the changelog. 

\* In my tests for log iterations in #5337, convergence behaviour is a lot *worse* (the internal Newton solve fails). I still need to figure out whether this is because of a bug or because the shape of the log-log curve (nearly straight-line segments with curved kinks in-between) is a problem for the Newton scheme. This might be fixed by using SUNDIALS.

**Comment by lhy11009** [comment]
@bobmyhill, thanks for your review. I am visiting family and slow to respond.
[ you see any improvement in convergence behaviour in any tests](https://github.com/geodynamics/aspect/pull/5338#:~:text=Do%20you%20see%20any%20improvement%20in%20convergence%20behaviour%20in%20any%20tests*%3F)

**Comment by lhy11009** [comment]
@bobmyhill I have tested the iteration history from the tests/diffusion_dislocation.prm. The results are positive.
[job_log_derivative.stdout.txt](https://github.com/geodynamics/aspect/files/12189465/job_log_derivative.stdout.txt)
[job_derivative.stdout.txt](https://github.com/geodynamics/aspect/files/12189484/job_derivative.stdout.txt)

File: source/material_model/diffusion_dislocation.cc [hunk 1]
```diff
@@ -277,7 +282,7 @@ namespace aspect
                              "Scaling coefficient for effective viscosity.");
 
           // Viscosity iteration parameters
-          prm.declare_entry ("Strain rate residual tolerance", "1e-22", Patterns::Double(0.),
+          prm.declare_entry ("Strain rate residual tolerance", "1e-10", Patterns::Double(0.),
                              "Tolerance for correct diffusion/dislocation strain rate ratio.");
```
Comment (bobmyhill) [inline source/material_model/diffusion_dislocation.cc:286]:
```suggestion
                             "Tolerance for determining the correct stress and viscosity from the strain rate by internal iteration. The tolerance is expressed as the difference between the natural logarithm of the input strain rate and the strain rate at the current iteration. This determines that strain rate is correctly partitioned between diffusion and dislocation creep assuming that both mechanisms experience the same stress.");
```

File: doc/modules/changes/20230718_lhy11009 [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Improved: The iteration scheme in the diffusion dislocation material model now uses the derivative of the logarithm of the strain rate to the logarithm of the stress.
```
Comment (bobmyhill) [inline doc/modules/changes/20230718_lhy11009:1]:
```suggestion
Improved: The iteration scheme in the diffusion dislocation material model now uses the derivative of the logarithm of the strain rate to the logarithm of the stress. This improves efficiency and stablility of the iterative scheme. The parameter "Strain rate residual tolerance" now corresponds to the residual for the logarithm of the strain rate (which close to the solution corresponds to the relative residual).
```

File: include/aspect/material_model/rheology/diffusion_creep.h [hunk 3]
```diff
@@ -114,6 +114,16 @@ namespace aspect
                                               const double temperature,
                                               const DiffusionCreepParameters creep_parameters) const;
 
+          /**
+           * Compute the logarithm of strain rate and first logarithm stress derivative as a function
```
Comment (bobmyhill) [inline include/aspect/material_model/rheology/diffusion_creep.h:118]:
```suggestion
           * Compute the logarithm of strain rate and first derivative with respect to
```

File: include/aspect/material_model/rheology/diffusion_creep.h [hunk 4]
```diff
@@ -114,6 +114,16 @@ namespace aspect
...
+           * of the logarithm stress based on the diffusion creep law.
```
Comment (bobmyhill) [inline include/aspect/material_model/rheology/diffusion_creep.h:119]:
```suggestion
           * the logarithm of the stress based on the diffusion creep law.
```

File: include/aspect/material_model/rheology/dislocation_creep.h [hunk 5]
```diff
@@ -111,6 +111,16 @@ namespace aspect
                                               const double temperature,
                                               const DislocationCreepParameters creep_parameters) const;
 
+          /**
+           * Compute the logarithm of strain rate and first logarithm stress derivative as a function
```
Comment (bobmyhill) [inline include/aspect/material_model/rheology/dislocation_creep.h:115]:
```suggestion
           * Compute the logarithm of strain rate and first derivative with respect to
```

File: include/aspect/material_model/rheology/dislocation_creep.h [hunk 6]
```diff
@@ -111,6 +111,16 @@ namespace aspect
...
+           * of the logarithm stress based on the dislocation creep law.
```
Comment (bobmyhill) [inline include/aspect/material_model/rheology/dislocation_creep.h:116]:
```suggestion
           * the logarithm of the stress based on the dislocation creep law.
```

**Review by bobmyhill — COMMENTED** [review]
Great! Thank you for the checks, they are really useful.

I have a few suggestions for improving docstrings and your changelog, but otherwise this looks excellent, thank you for doing this!

I'm away for a couple of weeks, but once those little changes are made I'm happy for this to be merged before I'm back.

**Review by bobmyhill — APPROVED** [review]

**Comment by bobmyhill** [comment]
Very nice, thank you!

**Pull request merged**